codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -358,18 +357,23 @@ public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, lon
      */
     @Override
     public ChannelFuture writeGoAway(ChannelHandlerContext ctx, int lastStreamId, long errorCode, ByteBuf debugData,
-            ChannelPromise promise) {
-        Http2Connection connection = connection();
-        if (connection.isGoAway()) {
+                                     ChannelPromise promise) {
+        try {
+            if (connection().goAwaySent() && connection().remote().lastKnownStream() < lastStreamId) {
+                throw connectionError(PROTOCOL_ERROR, "Last stream identifier must not increase between " +
+                                                      "sending multiple GOAWAY frames (was '%d', is '%d').",
+                                                      connection().remote().lastKnownStream(),
+                                                      lastStreamId);
+            }
+            connection().goAwaySent(lastStreamId);
```

Comment:
@nmittler - That sounds reasonable.  I am just considered with maintaining consistent state.  As long as the object is completely dead (reached termination state) then that is fine.  I was wresting with what to do on the RST_STREAM frame in this case...and I just punted and sent a GO_AWAY...assuming this would put the connection into a "termination state."